lib/src/license_detection/tokenizer.dart

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+part of 'license_detector.dart';
+
+/// A token in this context is a sequence of characters
+/// scanned from the original text separated by a space
+/// or new line.
+@visibleForTesting
+@sealed
+class Token {
+  /// Normalized form of the text in [span].
+  final String value;
+
+  /// Zero based position of token in a list of [Token]s obtained by tokenizing license text.
+  ///
+  /// License when tokenized through [tokenize] method yields a list of [Token].
+  /// The [index] denotes the offset (or index) of the token in this list.
+  final int index;
+
+  /// Location details of the token in original text.
+  final SourceSpan span;
+
+  @visibleForTesting
+  Token(this.value, this.index, this.span);
+
+  Token._fromSpan(this.span, this.index)
+      : value = _normalizeWord(span.text.toLowerCase());
+}
+
+/// Tokenizes license text and returns a list of [Token]s.
+///
+/// The tokens are scanned and normalized according to this procedure.
+/// 1. Tokens are words separated by spaces or new line. Ex - `hello new\nworld` -->[`hello`, `new`, `world`,].
+/// 2. Any standalone punctuations texts are ignored. Ex - `// % ^&^&*^ hello` --> [`hello`,]
+/// 3. Leading punctuations are ignored. Ex - `!hello $world&` --> [`hello`, `world&`,]
+/// 4. Tokens resembling list items are ignored. Ex - `1.`, `ix.`, `a)` .
+/// 5. Any tokens starting with alphabet are further cleaned to remove any
+///    non-alphabet characters. Ex - `(hello wo$r1ld` --> [`hello`, `world`,]
+/// 6. If a token starts with digit, any characters except `.`, `-` in the token value will be ignored.
+///    Dot at the end of the token will also be ignored. `1!@#$.1.1 1-1hj.23.` --> [`1.1.1`, `1-1.23`,].
+@visibleForTesting
+List<Token> tokenize(String text) {
+  final _scanner = SpanScanner(text);
+
+  var tokens = <Token>[];
+  var tokenID = 0;
+
+  Token? nextToken() {
+    if (_scanner.scan(_wordRegex)) {
+      return Token._fromSpan(_scanner.lastSpan!, tokenID++);
+    }
+
+    // Store new line tokens to identify and ignore tokens
+    // resembling list items while cleaning tokens.
+    if (_scanner.scan(_newLineRegex)) {
+      return Token('\n', tokenID++, _scanner.lastSpan!);
+    }
+
+    // Ignore whitespace
+    if (_scanner.scan(_horizontalWhiteSpaceRegex)) {
+      return null;
+    }
+
+    // Read only © and ignore other leading and standalone punctuation.
+    // if(_scanner.scan(RegExp(r'©'))){
+    //   return NewToken.fromSpan(_scanner.lastSpan!, tokenID++);
+    // }
+
+    // If none of the above conditions match, this implies
+    // the scanner is at standalone punctuation mark or leading
+    // punctuation in a word. Ignore them and move the scanner forward.
+    _scanner.readChar();
+    return null;
+  }
```

Comment:
I would. prefer `nextToken` to always return a token if possible.
Maybe something like:

```suggestion
  Token? nextToken() {
    while (!_scanner.isDone) {
      if (_scanner.scan(_wordRegex)) {
        return Token._fromSpan(_scanner.lastSpan!, tokenID++);
      }

      // Store new line tokens to identify and ignore tokens
      // resembling list items while cleaning tokens.
      if (_scanner.scan(_newLineRegex)) {
        return Token('\n', tokenID++, _scanner.lastSpan!);
      }

      // Ignore whitespace
      if (!_scanner.scan(_horizontalWhiteSpaceRegex))) {

        // If none of the above conditions match, this implies
        // the scanner is at standalone punctuation mark or leading
        // punctuation in a word. Ignore them and move the scanner forward.
        _scanner.readChar();
    }
    return null;
  }
```

Reply:
Yes, this seems much better, I'll make the changes.